src/common/meta/src/ddl/drop_database/cursor.rs

```diff
@@ -0,0 +1,141 @@
+// limitations under the License.
+
+use common_procedure::Status;
+use futures::TryStreamExt;
+use serde::{Deserialize, Serialize};
+use snafu::OptionExt;
+use table::metadata::TableId;
+
+use super::executor::DropDatabaseExecutor;
+use super::metadata::DropDatabaseRemoveMetadata;
+use super::DropTableTarget;
+use crate::ddl::drop_database::{DropDatabaseContext, State};
+use crate::ddl::DdlContext;
+use crate::error::{self, Result};
+use crate::key::table_route::TableRouteValue;
+use crate::key::DeserializedValueWithBytes;
+use crate::table_name::TableName;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DropDatabaseCursor {
+    target: DropTableTarget,
+}
+
+impl DropDatabaseCursor {
+    /// Returns a new [DropDatabaseCursor].
+    pub fn new(target: DropTableTarget) -> Self {
+        Self { target }
+    }
+
+    fn handle_reach_end(
+        &mut self,
+        ctx: &mut DropDatabaseContext,
+    ) -> Result<(Box<dyn State>, Status)> {
+        match self.target {
+            DropTableTarget::Logical => {
+                // Consumes the tables stream.
+                ctx.tables.take();
+
+                Ok((
+                    Box::new(DropDatabaseCursor::new(DropTableTarget::Physical)),
+                    Status::executing(true),
+                ))
+            }
+            DropTableTarget::Physical => Ok((
+                Box::new(DropDatabaseRemoveMetadata),
+                Status::executing(true),
+            )),
+        }
+    }
+
+    async fn handle_table(
+        &mut self,
+        ddl_ctx: &DdlContext,
+        ctx: &mut DropDatabaseContext,
+        table_name: String,
+        table_id: TableId,
+        table_route_value: DeserializedValueWithBytes<TableRouteValue>,
+    ) -> Result<(Box<dyn State>, Status)> {
+        match (self.target, table_route_value.get_inner_ref()) {
+            (DropTableTarget::Logical, TableRouteValue::Logical(_))
+            | (DropTableTarget::Physical, TableRouteValue::Physical(_)) => {
+                // TODO(weny): Maybe we can drop the table without fetching the `TableInfoValue`
+                let table_info_value = ddl_ctx
+                    .table_metadata_manager
+                    .table_info_manager()
+                    .get(table_id)
+                    .await?
+                    .context(error::TableNotFoundSnafu {
+                        table_name: &table_name,
+                    })?;
+                Ok((
+                    Box::new(DropDatabaseExecutor::new(
+                        TableName::new(&ctx.catalog, &ctx.schema, &table_name),
+                        table_id,
+                        table_info_value,
+                        table_route_value,
+                        self.target,
+                    )),
+                    Status::executing(true),
+                ))
+            }
+            _ => Ok((
+                Box::new(DropDatabaseCursor::new(self.target)),
```

Comment:
If I understand it correctly, you are trying to make the DropTableTarget met the current to-be-dropped table's route type, by retry fetching all the tables, right? But what if the first newly fetched table still not match the target, will it end up in a infinite loop?

Reply:
Then what about the current handling table? It's not dropped by DropDatabaseExecutor.

Reply:
Oh, I got you. The procedure will drop logical table first, then the physical table(the restriction of the metric engine). Finally, all tables will be dropped.